### PR TITLE
Bug 1222069 - [User Story] Re-arrange Pinned Pages

### PR DESCRIPTION
**Pull request by gmarty**

### CODE REVIEW

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=4074d1470fe020dd0f5851af17eb12b82174b90c

File: apps/homescreen/js/appsmetadata.js [hunk 1]
```diff
@@ -47,12 +52,12 @@
           }
 
           if (typeof entry.order !== 'undefined') {
-            txn.objectStore(DB_ORDER_STORE).
-              put({ id: entry.id, order: entry.order });
+            txn.objectStore(DB_ORDER_STORE)
+              .put({ id: entry.id, order: entry.order });
```
Comment (Cwiiis) [inline apps/homescreen/js/appsmetadata.js]: nit: personally I prefer to have the operator at the end of the broken line rather than the beginning of the new line (I find it better for readability).

File: apps/homescreen/js/appsmetadata.js [hunk 2]
```diff
@@ -47,12 +52,12 @@
...
           }
           if (typeof entry.icon !== 'undefined') {
-            txn.objectStore(DB_ICON_STORE).
-              put({ id: entry.id, icon: entry.icon });
+            txn.objectStore(DB_ICON_STORE)
+              .put({ id: entry.id, icon: entry.icon });
```
Comment (Cwiiis) [inline apps/homescreen/js/appsmetadata.js]: Same here.

File: apps/homescreen/style/pages-panel.css [hunk 3]
```diff
@@ -105,7 +109,7 @@ div#boxes > div:nth-child(4) {
   height: 16.5rem;
   padding: 0 0 1.5rem 1.5rem;
   outline: 0;
-  transition: opacity 0.2s, transform 0.2s;
+  transition: transform 0.2s, opacity 0.2s;
```
Comment (Cwiiis) [inline apps/homescreen/style/pages-panel.css]: nit: unnecessary change?

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=05ab154c366a7bcc3105206ede36d2f89e8f027c

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=cd349985b6b2a091672ffee364ad22cd740a7062

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=67cd3eb02d11d91d61a6cc325446782d83cab43e

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=74039dece75bfda64d02c824588fd87df89ca16e

File: apps/homescreen/js/pages.js [hunk 4]
```diff
@@ -1,37 +1,77 @@
...
+    this.draggedIndex = -1;
+    this.autoScrollInterval = null;
+    this.autoScrollOverflowTimeout = null;
+    this.hoverIcon = null;
+
     // Edit mode
     this.editMode = false;
 
-    // Dialogs
-    this.dialogs = [];
```
Comment (Cwiiis) [inline apps/homescreen/js/pages.js]: I'd rather this change wasn't made right now - we have a bug open to add a confirm dialog for deleting pinned pages and we'll just have to add this back in again (but it could easily be forgotten).

File: apps/homescreen/js/pages.js [hunk 5]
```diff
@@ -43,79 +83,136 @@
...
+      this.storePagesOrder();
+      this.refreshGridSize();
+      this.snapScrollPosition();
     });
   }
 
   Pages.prototype = {
     updatePinnedPage: function(card, page) {
+      if (this.startupMetadata !== null) {
+        var entry = this.startupMetadata.findIndex(data => data.id === page.id);
```
Comment (Cwiiis) [inline apps/homescreen/js/pages.js]: Great use of fat-arrow functions :)

File: apps/homescreen/js/pages.js [hunk 6]
```diff
@@ -260,9 +488,135 @@
         }
         break;
 
+      // Disable scrolling during dragging, and display bottom-bar
+      case 'drag-start':
+        console.debug('Drag-start on ' + e.detail.target.dataset.id);
```
Comment (Cwiiis) [inline apps/homescreen/js/pages.js]: Let's remove these console.debug calls - the ones in apps.js were added to try to debug a bug that has since been marked as WORKSFORME.

File: apps/homescreen/js/window.js [hunk 7]
```diff
@@ -123,7 +123,7 @@
         }, HASH_CHANGE_DEBOUNCE);
 
         // If a dialog is showing, cancel the dialog
-        for (var dialog of this.apps.dialogs.concat(this.pages.dialogs)) {
+        for (var dialog of this.apps.dialogs) {
```
Comment (Cwiiis) [inline apps/homescreen/js/window.js]: (see above)

File: apps/homescreen/test/unit/window_test.js [hunk 8]
```diff
@@ -271,16 +271,6 @@ suite('AppWindow', () => {
         appWindow.apps.dialogs = realDialogs;
       });
 
-      test('should cancel pages-panel dialogs', done => {
```
Comment (Cwiiis) [inline apps/homescreen/test/unit/window_test.js]: See above.

File: build/csslint/xfail.list [hunk 9]
```diff
@@ -58,6 +58,7 @@ apps/homescreen/bower_components/base/base.css 4 3
 apps/homescreen/bower_components/gaia-fonts/style.css 0 1
 apps/homescreen/bower_components/gaia-theme/gaia-theme.css 5 0
 apps/homescreen/style/apps-panel.css 1 0
+apps/homescreen/style/pages-panel.css 1 0
```
Comment (Cwiiis) [inline build/csslint/xfail.list]: I assume this is still necessary? (i.e. we haven't upgraded/fixed csslint)

File: apps/homescreen/js/pages.js [hunk 10]
```diff
@@ -260,9 +488,135 @@
...
+        this.dragging = true;
+        this.shouldEnterEditMode = true;
+        document.body.classList.add('dragging');
+        this.scrollable.style.overflow = 'hidden';
+        this.draggedIndex = this.getChildIndex(e.detail.target);
+        break;
+
+      case 'drag-finish':
+        console.debug('Drag-finish');
```
Comment (Cwiiis) [inline apps/homescreen/js/pages.js]: and here

File: apps/homescreen/js/pages.js [hunk 11]
```diff
@@ -260,9 +488,135 @@
...
+          this.hoverIcon = null;
+        }
+
+        // Restore normal drag-and-drop after dragging selected icons
+        this.pages.dragAndDropTimeout = -1;
+        break;
+
+      // Handle app/site editing and dragging to the end of the icon grid.
+      case 'drag-end':
+        console.debug('Drag-end, target: ' + (e.detail.dropTarget ?
```
Comment (Cwiiis) [inline apps/homescreen/js/pages.js]: and here

File: apps/homescreen/js/pages.js [hunk 12]
```diff
@@ -260,9 +488,135 @@
...
+            e.detail.dropTarget.dataset.id : 'none'));
+        if (e.detail.dropTarget === null &&
+          e.detail.clientX >= this.iconsLeft &&
+          e.detail.clientX < this.iconsRight) {
+          // If the drop target is null, and the client coordinates are
+          // within the panel, we must be dropping over the start or end of
+          // the container.
+          e.preventDefault();
+          var bottom = e.detail.clientY < this.lastWindowHeight / 2;
+          console.debug('Reordering dragged icon to ' +
```
Comment (Cwiiis) [inline apps/homescreen/js/pages.js]: and here

File: apps/homescreen/js/pages.js [hunk 13]
```diff
@@ -260,9 +488,135 @@
...
+          if (this.editMode || this.shouldEnterEditMode) {
+            e.preventDefault();
+            this.enterEditMode(icon);
+          }
+        }
+        break;
+
+      // Save the app grid after rearrangement
+      case 'drag-rearrange':
+        console.debug('Drag rearrange');
```
Comment (Cwiiis) [inline apps/homescreen/js/pages.js]: and here

File: build/csslint/xfail.list (hunk 9 above)
Comment (gmarty) [inline build/csslint/xfail.list]: Yes, unfortunately the properties related to snap positioning are still not supported.
